### ISSUE TITLE
NaN in onSlideChanged

### ISSUE THREAD

**Issue by dlewak** [iosscripts/iosslider#157]
I have a callback to onSlideChanged and I sometimes see currentSlideNumber with a value of NaN.  

The setup is this:  I have a slider with 32 slides.  The slider is setup to display only 3 slides at a time.  Each slide has an effective width of 150 pixels, and the slider parent has a width of 464 pixels.

I did some minimal debugging.  The callstack is

  onSlideChanged
  slowScrollHorizontalInterval
  slowScrollHorizontalIntervalTimer
  changeSlide

The NaN value originates from line 734 in v1.34:   

  var endOffset = childrenOffsets[slide]; 

In this case, slide is 31, but childrenOffsets only has 30 entries.

I tracked the population of childrenOffsets down, and found that in the init function, the line 1221

  childrenOffsets.splice(lastChildOffset + 1, childrenOffsets.length);

truncates childrenOffsets down from 32 entries to 30 entries, causing the later NaN issue.

**Comment by iosscripts** [comment]
Just to be sure that you are doing everything right. Does your onSlideChange function looks something like this?

``` js
function onSlideChangeCallback() {

    /* accessing the current slide number like so */
    var currentSlideNumber = $('.iosSlider').data('args').currentSlideNumber;

}
```

If so, can you show me what your callback looks like?

**Comment by dlewak** [comment]
My function looks like this:

``` js
   onSlideChanged: function(args) {
         var currentSlideNumber = args.currentSlideNumber;
         var c = $('.iosSlider').data('args').currentSlideNumber;
   },
```

both currentSlideNumber and c have a NaN value under the condition that there are 32 to slides and 

``` js
self.$('.iosSlider').iosSlider('goToSlide', 32);
```

is called. If it matters, the callback is set in the slider initialization, which happens during each backbone view render:

``` js
   init_slider: function () {
      this.$('.iosSlider').iosSlider({
         snapToChildren: true,
         startAtSlide: this.restoreViewPort ? this.currentSlideNumber : 1,
         scrollbar: false,
         desktopClickDrag: true,
         onSliderLoaded: this.onSlideChanged,
         onSlideChange: this.onSlideChanged,
         onSliderUpdate: this.onSlideChanged
      });
   },
   render: function() {
      var self = this;
      var context = {
         listName: this.options.listName
      };
      dust.render('explorer_list', context, function(err, out) {
         if (err) {
            dispatcher.trigger('errorMessage', err);
            return;
         }
         self.$el.html(out);

         // render slides
         self.$(".iosSlider").iosSlider('destroy');
         self.$('.slider').empty();
         _.each(self.collection.models, function(pkg) {
            self.$('.slider').append(self.explorerItemViews[pkg.cid].render().$el);
            self.explorerItemViews[pkg.cid].delegateEvents();
         });
         self.init_slider();
      });
      return this;
   },
```

Both currentSlideNumber and c have a NaN value under the condition that

**Comment by iosscripts** [comment]
``` js
var currentSlideNumber = $('.iosSlider').data('args').currentSlideNumber;
```

Above is the correct syntax for accessing the current slide number on slide change or anytime in your JS actually. Right after you initialize (self.init_slider();), if you dump $('.iosSlider').data('args'), do you see any data? or is it undefined?

**Comment by dlewak** [comment]
$(‘.iosSlider’).data(‘args’) is undefined the first time the slider it initialized.    Every successive render destroys the slider and recreates it, and for those render calls I do see data.

**Comment by iosscripts** [comment]
I have applied a fix to v1.1.38 which addresses the undefined error when navigating to slides > childrenOffset when using goToSlide. I believe this will solve your issue. Let me know if it works for you.

**Comment by dlewak** [comment]
This unfortunately doesn’t work.

For Line 755:
var endOffset = childrenOffsets[slide];
endOffset is NaN,  childrenOffsets has 30 elements, slide at this point is 30.   The slider should have 32 elements, I show 3 elements at a time, and in this case I’m trying to go to the last slide, 32.

Should I be indexing starting from 0 or 1?  Your documentation says the startAtSlide default value is 1, which implies indexing from 1.  If I index from 0 when I call goToSlide, I don’t get a NaN (with v.1.1.38.  I did get a NaN with previous versions).

A new problem is that the padding between slides is now 0, it was not 0 previously.

**Comment by iosscripts** [comment]
childrenOffsets is correct. There are 30 different offsets which can be snapped to (that is, the left side of the slide can hit the left edge of the iosSlider element). The last 2 offsets being unavailable at the end of the slider because you cannot slide directly to them because the end of the slider has been hit.

So with my fix from v1.1.38, using .iosSlider('gotoSlide', 32); should take you to slide 30. (30, 31, 32) all go to the same spot because the end of the slider is hit. You should not be able to slide all the way to slide 32.

To recap. If .iosSlider('gotoSlide', 32); is taking you to the very end of the slider, the script is behaving correctly. If you are getting a NaN error, that means I haven't fixed it.

Does that make sense? or am I not understanding your issue still :S? If I could see your implementation that would help.

> A new problem is that the padding between slides is now 0, it was not 0 previously.

Shit. So your CSS is being overridden correct? I will fix it.

**Comment by dlewak** [comment]
Yes, that makes sense.  I get a NaN error for .iosSlider('gotoSlide', 32);, so you haven’t fixed the bug.

**Comment by iosscripts** [comment]
v1.1.39 has a fix for the margin/padding issue.

Where is the NaN error? In the .data('args').currentSlideNumber still?

If that's the case, I cannot reproduce it in v1.1.38. I have created a demo triggering .iosSlider('gotoSlide', 1000); and it animates to the last slide correctly. Here is my code:

``` html
<div class = 'iosSlider'>

    <div class = 'slider'>

        <div class = 'item item1'></div>

        <div class = 'item item2'></div>

        …

        <div class = 'item item31'></div>

        <div class = 'item item32'></div>

    </div>

</div>
```

``` css
.iosSlider {
    position: relative;
    top: 0;
    left: 0;
    overflow: hidden;
    width: 464px;
    height: 150px;
}

.iosSlider .slider {
    width: 100%;
    height: 100%;
}

.iosSlider .slider .item {
    position: relative;
    top: 0;
    left: 0;
    width: 150px;
    height: 150px;
    margin: 0;
}
```

``` js
$(document).ready(function() {

    $('.iosSlider').iosSlider({
        snapToChildren: true,
        desktopClickDrag: true,
        onSlideChange: slideChange
    });

    /* callback */
    function slideChange() {
        console.log($('.iosSlider').data('args'));
    }

    /* output data after initialization */
    console.log($('.iosSlider').data('args'));

    /* attempt to navigate to slide 1000 */
    setTimeout(function() {
        $('.iosSlider').iosSlider('goToSlide', 1000);
    }, 3000);

});
```

In my case, childrenOffsets.length = 29 and the number of slides is 32. When I call goToSlide: 1000 in the setTimeout, it moves to the last slide (slide 29). No NaN's are returned for currentSlideNumber in any of the 30 slideChange() logging or the output after initialization.

What should I change to reproduce your issue?

**Comment by dlewak** [comment]
The code below produces this output:

Object { settings={...}, data={...}, slideChanged=false, more...}
number is NaN

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>

<script src="js/jquery.js" type="text/javascript"></script>

<script src='js/jquery.iosslider.js' type='text/javascript'></script>

<style type="text/css">
.iosSlider {
    position: relative;
    top: 0;
    left: 0;
    overflow: hidden;
    width: 464px;
    height: 150px;
}
.iosSlider .slider {
    width: 100%;
    height: 100%;
}
.iosSlider .slider .item {
    position: relative;
    top: 0;
    left: 0;
    width: 150px;
    height: 150px;
    margin: 0;
}
</style>

</head>
<body>

<div class = 'iosSlider'>
    <div class = 'slider'>
        <div class = 'item item1'>| 1 |</div>
        <div class = 'item item2'>| 2 |</div>
        <div class = 'item item3'>| 3 |</div>
        <div class = 'item item4'>| 4 |</div>
        <div class = 'item item5'>| 5 |</div>
        <div class = 'item item6'>| 6 |</div>
        <div class = 'item item7'>| 7 |</div>
        <div class = 'item item8'>| 8 |</div>
        <div class = 'item item9'>| 9 |</div>
        <div class = 'item item10'>| 10 |</div>
        <div class = 'item item11'>| 11 |</div>
        <div class = 'item item12'>| 12 |</div>
        <div class = 'item item13'>| 13 |</div>
        <div class = 'item item14'>| 14 |</div>
        <div class = 'item item15'>| 15 |</div>
        <div class = 'item item16'>| 16 |</div>
        <div class = 'item item17'>| 17 |</div>
        <div class = 'item item18'>| 18 |</div>
        <div class = 'item item19'>| 19 |</div>
        <div class = 'item item20'>| 20 |</div>
        <div class = 'item item21'>| 21 |</div>
        <div class = 'item item22'>| 22 |</div>
        <div class = 'item item23'>| 23 |</div>
        <div class = 'item item24'>| 24 |</div>
        <div class = 'item item25'>| 25 |</div>
        <div class = 'item item26'>| 26 |</div>
        <div class = 'item item27'>| 27 |</div>
        <div class = 'item item28'>| 28 |</div>
        <div class = 'item item29'>| 29 |</div>
        <div class = 'item item30'>| 30 |</div>
        <div class = 'item item31'>| 31 |</div>
        <div class = 'item item32'>| 32 |</div>
    </div>
</div>

<script type="text/javascript">
$(document).ready(function() {
    $('.iosSlider').iosSlider({
        snapToChildren: true,
        desktopClickDrag: true,
        onSlideChange: slideChange
    });
    /* callback */
    function slideChange() {
        var number = $('.iosSlider').data('args').currentSlideNumber;
        if (isNaN(number)) {
            console.log("number is NaN");
        } else {
            console.log("number="+number);
        }
    }
    /* output data after initialization */
    console.log($('.iosSlider').data('args'));
    /* attempt to navigate to slide 1000 */
    setTimeout(function() {
        $('.iosSlider').iosSlider('goToSlide', 32);
    }, 3000);
});
</script>

</body>
</html>

**Comment by iosscripts** [comment]
Ok. I was finally able to reproduce it. Thanks for the code. Please download v1.1.40 for the fix.

**Comment by dlewak** [comment]
Almost there. It works for 32, but not 31. I think you need to change

slide = ((slide - 1) > data.childrenOffsets.length) ? data.childrenOffsets.length - 1 : slide - 1;

to

slide = ((slide - 1) > data.childrenOffsets.length - 1) ? data.childrenOffsets.length - 1 : slide - 1;

**Comment by marcwhitbread** [comment]
We could just drop both the -1's then:

slide = (slide > data.childrenOffsets.length) ? data.childrenOffsets.length - 1 : slide - 1;

I will double check with goToSlide: 31/32.
